File: stuff_usecases/src/main/java/android/montreal/devfest/stuff/usescases/ManageRandomStuffUseCaseImpl.java
package android.montreal.devfest.stuff.usescases;

import android.montreal.devfest.policies.RandomStuffVerifier;

class ManageRandomStuffUseCaseImpl implements ManageRandomStuffUseCase {
    private final RandomStuffGetter getter;
    private final RandomStuffSetter setter;
    private final RandomStuffVerifier verifier;

    public ManageRandomStuffUseCaseImpl(RandomStuffVerifier verifier, RandomStuffSetter setter, RandomStuffGetter getter) {
        this.verifier = verifier;
        this.getter = getter;
        this.setter = setter;
    }

    @Override
    public RandomStuffData getRandomStuff() throws GetRandomStuffException {
        final RandomStuffData randomStuff = this.getter.getRandomStuff();

        if (!this.verifier.isValid(DataToEntityMapper.map(randomStuff))) {
            throw new GetRandomStuffException("No random stuff here!");
        }

        return randomStuff;
    }

    @Override
    public void setRandomStuff(RandomStuffData randomStuff) throws SetRandomStuffException {
        if (!this.verifier.isValid(DataToEntityMapper.map(randomStuff))) {
            throw new SetRandomStuffException("Can't use this random stuff!");
        }

        this.setter.setRandomStuff(randomStuff);
    }
}

File: stuff_usecases/src/main/java/android/montreal/devfest/stuff/usecases/ManageRandomStuffUseCaseImpl.kt
package android.montreal.devfest.stuff.usecases

import android.montreal.devfest.stuff.policies.RandomStuffVerifier

internal class ManageRandomStuffUseCaseImpl(private val verifier: RandomStuffVerifier, private val setter: RandomStuffSetter, private val getter: RandomStuffGetter) : ManageRandomStuffUseCase {

    @Throws(GetRandomStuffException::class)
    override fun getRandomStuff(): RandomStuffData {
        val randomStuff = this.getter.getRandomStuff()

        if (!this.verifier.isValid(DataToEntityMapper.map(randomStuff)!!)) {
            throw GetRandomStuffException("No random stuff here!")
        }

        return randomStuff
    }

    @Throws(SetRandomStuffException::class)
    override fun setRandomStuff(stuffData: RandomStuffData) {
        if (!this.verifier.isValid(DataToEntityMapper.map(stuffData)!!)) {
            throw SetRandomStuffException("Can't use this random stuff!")
        }
        this.setter.setRandomStuff(stuffData)
    }
}
